anchore_engine/util/docker.py

```diff
@@ -149,3 +151,238 @@ def parse_dockerimage_string(instr, strict=True):
+
+        :return:
+        """
+        layers = [layer["blobSum"] for layer in self.raw.get("fsLayers", [])]
+
+        # Reverse the order to normalize order w/V2
+        layers.reverse()
+
+        return layers
+
+    def _history(self):
+        history = []
+        count = 0
+
+        layers = self._layer_ids()
+        layers.reverse()  # Reverse again to get back to original order
+
+        for layer in self.raw.get("history", []):
+            hel = json.loads(layer["v1Compatibility"])
+            lsize = hel.get("Size", 0)
+
+            lcreatedby = ""
+            cmds = hel.get("container_config", {}).get("Cmd", [])
+            if cmds:
+                lcreatedby = " ".join(cmds)
+
+            lcreated = hel.get("created", "")
+
+            lid = layers[count]
+            count = count + 1
+            history.append(
+                {
+                    "Created": lcreated,
+                    "CreatedBy": lcreatedby,
+                    "Comment": "",
+                    "Id": lid,
+                    "Size": lsize,
+                    "Tags": [],
+                }
+            )
+
+        return history
+
+    def _infer_dockerfile(self):
+        # get dockerfile_contents (translate history to guessed DF)
+        dockerfile_contents = "FROM scratch\n"
+        for hel in self._history():
+            patt = re.match(r"^/bin/sh -c #\(nop\) +(.*)", hel["CreatedBy"])
+            if patt:
+                cmd = patt.group(1)
+            elif hel["CreatedBy"]:
+                cmd = "RUN " + hel["CreatedBy"]
+            else:
+                cmd = None
+            if cmd:
+                dockerfile_contents = dockerfile_contents + cmd + "\n"
+
+        return dockerfile_contents
+
+
+class DockerV2ManifestMetadata:
+    """
+    Processed manifest data that exposes the history, inferred dockerfile, etc based on a Docker Image Manifest v2 schema 2: https://docs.docker.com/registry/spec/manifest-v2-2/
+    """
+
+    def __init__(self, manifest_json, image_config_json):
+        """
+
+        :param manifest_json:
+        :param image_config_json:
+        :param downloaded_blob_list: list of string filenames for blobs downloaded, assumed to be sh256 digests
+        """
+
+        self.raw = manifest_json
+        self.image_config = image_config_json
+        self.layer_ids = self._layer_ids()
+        self.history = self._history()
+        self.inferred_dockerfile = self._infer_dockerfile()
+        self.architecture = self._architecture()
+
+    def _architecture(self):
+        if not self.image_config:
+            # Cannot get it without the config, manifest is insufficient
+            return None
+
+        return self.image_config.get("architecture")
+
+    def _layer_ids(self):
+        # Filter out all empty layers
+        return [
+            x.get("digest")
+            for x in self.raw.get("layers")
+            if not x.get("empty_layer", False)
```

Comment:
Doesn't look like this needs to fallback to a `False` if it is empty does it? it would still evaluate the same?

Reply:
the field is a boolean, so added this for consistency rather than True or None, but the effect is the same.